### DISCUSSION TITLE
Issues with string interpolation

### DISCUSSION THREAD

**Discussion by johnmoxley** [savonet/liquidsoap#2220] [Q&A] [answered]
Liquidsoap v2.0.2 via Docker

I have a script where I'm trying to run a simple archiver, which will create files for 30 minute chunks of audio. The intended format of the output file would be `DayHHMM` (e.g. `Mon1300` or `Tue0230`).

I'm using `output.file` with the `reopen_when` option to break up the files into chunks.

Here's a simplified version of my script below:

```
output_base_path = '/archive/test/'
output_day = { (1w)?"Mon":((2w)?"Tue":((3w)?"Wed":((4w)?"Thu":((5w)?"Fri":((6w)?"Sat":((7w)?"Sun":"Sun")))))) }
output_min = { (00m00s-29m59s)?"00":"30" }

# Having issues here
output_filename = '$(day)%H$(min)__%H%M%S.mp3'

output.file(%mp3,
        output_base_path ^ output_filename % [("day",output_day()),("min",output_min())],
        input,
        reopen_when={0m or 30m})
```

In the script, I've also prepended `__%H%M%S` to the filename for testing purposes. 

However, when `reopen_when` executes, it doesn't re-evaluate the `output_day` and `output_min` functions, but it *will* re-evaluate the `%H` and others.

This results in filenames like:
 - `Sun2300__230732.mp3` (Script started at 23:07, filename correct)
 - `Sun2300__233000.mp3` (`reopen_when` activates, the `$(min)` part of the string has not been re-evaluated)

Is there a better way to go about this? Thanks for giving this a look, I appreciate any pointers.

**Comment by toots** [comment, marked as answer]
Thanks for your question. String interpolation is outdated and should be replaced by proper string getters now. I have added a ticket to fully remove it from the code base here: https://github.com/savonet/liquidsoap/issues/2221

To answer your question here, the issues comes from the fact that this:
```ruby
output.file(%mp3,
        output_base_path ^ output_filename % [("day",output_day()),("min",output_min())],
        input,
        reopen_when={0m or 30m})
```
Means that the line `output_base_path ^ output_filename % [("day",output_day()),("min",output_min())]` is executed only once, when the script is initially called. It is never executed again and, so, the suffix remains the same.

The solution is to make the whole thing a string getter, i.e. a function of type `() -> string`. Something like:
```ruby
output_base_path = "/archive/test/"
day_mappings = [
  (0, "Sun"),
  (1, "Mon"),
  (2, "Tue"),
  (3, "Wed"),
  (4, "Thu"),
  (5, "Fri"),
  (6, "Sat")
]

def output_file() =
  # With 2.1.x: let {hour, min, sec, week_day} = time.local()
  current_time = time.local()
  hour = current_time.hour
  min = current_time.min
  sec = current_time.sec
  week_day = current_time.week_day

  output_min = min < 30 ? "00" : "30"
  output_day = list.assoc(week_day, day_mappings)

  "#{output_base_path}#{output_day}#{hour}#{output_min}_#{hour}#{min}#{sec}.mp3"
end

output.file(%mp3,
        output_file,
        input,
        reopen_when={0m or 30m})
```
This is based off the ongoing `2.0.3-preview` branch, which will come out soon as the next stable release.

**Comment by johnmoxley** [comment]
Thank you for the detailed reply

**Comment by theotherjenkutler** [comment]
@toots Is there a version of this that works with 2.0.2? When I use current_time = time.local() it doesn't appear to return a callback. I get an error that current_time has no method for year, month, day, hour, etc.

I should also note that the 2.1.x solution throws a syntax error.

**Reply by toots** [reply]
Hi! `current_time` shouldn't be a callback but a record containing all the relevant information. The name of the methods has changed in `2.0.3` to make it more user friend. The doc for the `2.0.2` operator is here: https://www.liquidsoap.info/doc-2.0.2/reference.html#time.local
